ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs

```diff
@@ -472,106 +476,143 @@ mkInternalStateDB :: (HasCallStack, MonadThrow m, MonadCatch m, Ord blockId)
+mkInternalStateDB hasFS@HasFS{..} err parser maxBlocksPerFile =
+  wrapFsError hasFsErr err $ do
+    createDirectoryIfMissing True dir
+    allFiles <- Set.map toFsPath <$> listDirectory dir
     mkInternalState hasFS err parser maxBlocksPerFile allFiles
   where
     dir = mkFsPath []
 
     toFsPath :: String -> FsPath
     toFsPath file = mkFsPath [file]
 
+-- | This function may create a new
 mkInternalState :: forall blockId m h e. (HasCallStack, MonadCatch m, Ord blockId)
                 => HasFS m h
                 -> ErrorHandling (VolatileDBError blockId) m
                 -> Parser e m blockId
                 -> Int
                 -> Set FsPath
                 -> m (InternalState blockId h)
-mkInternalState hasFS@HasFS{..} err parser n files = wrapFsError hasFsErr err $ do
-    lastFd <- findNextFd err files
-    let
+mkInternalState hasFS err parser n files =
+  wrapFsError (hasFsErr hasFS) err $ do
+    lastFd <- findLastFdM err files
+    mkInternalState' hasFS err parser n files lastFd
+
+mkInternalState' :: forall blockId m h e. (HasCallStack, MonadCatch m, Ord blockId)
+                 => HasFS m h
+                 -> ErrorHandling (VolatileDBError blockId) m
+                 -> Parser e m blockId
+                 -> Int
+                 -> Set FsPath
+                 -> Maybe FileId
+                 -> m (InternalState blockId h)
+mkInternalState' hasFS err parser n files lastFd =
+    go Index.empty Map.empty Map.empty Nothing [] (Set.toList files)
+    where
+        newFileInfo mp newIndex =
+            ( newFile
+            , newIndex + (1 :: Int)
+            , Index.insert newFile FileInfo.empty mp
+            , 0 )
+            where
+                newFile = filePath newIndex
+
+        errorMsg = concat
+            [ "Volatile db invariant violation:"
+            , "A file was found with less than "
+            , show n
+            , " blocks, but there are no files parsed."]
+
+        truncateOnError Nothing _ _ = return ()
+        truncateOnError (Just _) file offset =
+            -- the handle of the parser is closed at this point. We need
+            -- to reOpen the file in AppendMode now (parser opens with
+            -- ReadMode).
+            --
+            -- Note that no file is open at this point, so we can safely
+            -- open with AppendMode any file, without the fear of opening
+            -- multiple concurrent writers, which is not allowed.
+            --
+            withFile hasFS file (AppendMode AllowExisting) $ \hndl ->
+                hTruncate hasFS hndl (fromIntegral offset)
+
         go :: Index blockId
            -> ReverseIndex blockId
            -> SuccessorsIndex blockId
            -> Maybe (blockId, SlotNo)
-           -> [(FileId, FsPath, FileSize)] -- The relative path and size of the files with less than n blocks, if any found already.
+           -> [(FileId, FsPath, FileSize)] -- Info of files with < n blocks.
```

Comment:
Even though it is a local function, I would make it a Haddock docstring (`-- ^ ..`).